compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -193,10 +193,22 @@ object Applications {
         productSelectorTypes(unapplyResult, pos)
           // this will cause a "wrong number of arguments in pattern" error later on,
           // which is better than the message in `fail`.
+      else if unapplyResult.derivesFrom(defn.NonEmptyTupleClass) then
+        foldApplyTupleType(unapplyResult)
       else fail
     }
   }
 
+  def foldApplyTupleType(tp: Type)(using Context): List[Type] =
+    object tupleFold extends TypeAccumulator[List[Type]]:
+      override def apply(accum: List[Type], t: Type): List[Type] =
+        t match
+          case AppliedType(tycon, x :: x2 :: Nil) if tycon.typeSymbol == defn.PairClass =>
+            apply(x :: accum, x2)
+          case x => foldOver(accum, x)
+    end tupleFold
+    tupleFold(Nil, tp).reverse
```

Comment:
I think I overshot by thinking we should do a fold here.  I think that's the cause of the failure.  We should just traverse the type structure and do the various unwrapping calls ourselves, e.g dealiasing, unwrapping annotations, etc.

Reply:
It looks like the fold does the right thing here, the problem is that we didn't special-case for `unapplySeq`. I added a commit to do so, let's see if it breaks anything else.

Reply:
Oh, actually it might not be the fold.  But it's nothing to do with unapplySeq.  We just didn't carry over the `&& !isUnapplySeq` that was also present for the product branch.

Moving it has the same affect (though it keeps it separate from the product match).  But that unapplySeq change isn't right.

Reply:
Why not? Shouldn't we also support generic tuples as a return type of `unapplySeq`, and shouldn't this be special-cased?

Reply:
Fair enough.  But if you want to support it you should add tests that it actually works.